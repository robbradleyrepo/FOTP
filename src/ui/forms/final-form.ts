import { FieldErrors, FormErrors } from "@sss/forms";
import {
  FieldInputProps,
  FieldMetaState,
  FormProps as FinalFormProps,
  FormRenderProps as FinalFormRenderProps,
} from "react-final-form";

export interface FieldRenderProps<
  FieldValue = string,
  T extends HTMLElement = HTMLElement
> {
  input: FieldInputProps<FieldValue, T>;
  meta: Omit<FieldMetaState<FieldValue>, "error" | "submitError"> & {
    error?: FieldErrors;
    submitError?: FieldErrors;
  };
}

export type FormProps = Omit<FinalFormProps, "onSubmit"> & {
  onSubmit: (
    ...args: Parameters<FinalFormProps["onSubmit"]>
  ) => Promise<FormErrors | undefined | void> | FormErrors | undefined | void;
};

export type FormRenderProps = Omit<
  FinalFormRenderProps,
  "error" | "errors" | "submitError" | "submitErrors"
> & {
  error?: FieldErrors;
  errors?: FormErrors;
  submitError?: FieldErrors;
  submitErrors?: FormErrors;
};
