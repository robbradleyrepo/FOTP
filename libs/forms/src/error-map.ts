import { UserError } from "@sss/ecommerce/common";
import { FORM_ERROR } from "final-form";

import { FormErrors } from "./types";

export interface ErrorMap {
  [key: string]: ErrorMap | string;
}

const findErrorKey = (
  field: string[] | null,
  errorMap: ErrorMap
): string | undefined => {
  if (field === null) {
    return FORM_ERROR;
  }

  const [first, ...rest] = field;
  const branch = errorMap[first];

  if (branch) {
    return typeof branch === "string" ? branch : findErrorKey(rest, branch);
  }
};

export const getUserErrorsMapper = (errorMap: ErrorMap) => (
  userErrors: UserError[]
) => {
  const errors: FormErrors = {};
  const unmappedErrors: UserError[] = [];

  userErrors.forEach((error: UserError) => {
    const { field, message } = error;
    const key = findErrorKey(field, errorMap);

    if (!key) {
      unmappedErrors.push(error);

      return;
    }

    const currentMessages = errors[key] || [];

    errors[key] = [...currentMessages, message];
  });

  if (unmappedErrors.length) {
    throw new Error(
      `Customer user errors contained unmapped fields:\n  ${unmappedErrors.map(
        (error) => `${JSON.stringify(error)}\n`
      )}`
    );
  }

  return errors;
};
