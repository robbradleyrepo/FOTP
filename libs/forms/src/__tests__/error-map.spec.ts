import { FORM_ERROR } from "final-form";

import { getUserErrorsMapper } from "../error-map";

describe("getUserErrorsMapper", () => {
  it("should handle generic errors", () => {
    expect(
      getUserErrorsMapper({})([
        {
          code: "CUSTOMER_DISABLED",
          field: null,
          message:
            "We have sent an email to developers@fotp.com, please click the link included to verify your email address.",
        },
      ])
    ).toEqual({
      [FORM_ERROR]: [
        "We have sent an email to developers@fotp.com, please click the link included to verify your email address.",
      ],
    });
  });

  it("should handle errors that apply to specific fields", () => {
    expect(
      getUserErrorsMapper({ input: { firstName: "bar" } })([
        {
          code: "CONTAINS_HTML_TAGS",
          field: ["input", "firstName"],
          message: "First name cannot contain HTML tags",
        },
      ])
    ).toEqual({
      bar: ["First name cannot contain HTML tags"],
    });
  });

  it("should throw if any errors are missing a map entry", () => {
    expect(() =>
      getUserErrorsMapper({
        input: { password: "password" },
      })([
        {
          code: "CUSTOMER_DISABLED",
          field: null,
          message:
            "We have sent an email to developers@fotp.com, please click the link included to verify your email address.",
        },
        {
          code: "CONTAINS_HTML_TAGS",
          field: ["input", "firstName"],
          message: "First name cannot contain HTML tags",
        },
        {
          code: "CONTAINS_HTML_TAGS",
          field: ["input", "lastName"],
          message: "Last name cannot contain HTML tags",
        },
        {
          code: "PASSWORD_STARTS_OR_ENDS_WITH_WHITESPACE",
          field: ["input", "password"],
          message: "Password starts or ends with spaces.",
        },
      ])
    ).toThrowErrorMatchingSnapshot();
  });

  it("should handle multiple overlapping errors", () => {
    expect(
      getUserErrorsMapper({ input: { password: "password" } })([
        {
          code: "TOO_SHORT",
          field: ["input", "password"],
          message: "Password is too short (minimum is 5 characters)",
        },
        {
          code: "PASSWORD_STARTS_OR_ENDS_WITH_WHITESPACE",
          field: ["input", "password"],
          message: "Password starts or ends with spaces.",
        },
      ])
    ).toEqual({
      password: [
        "Password is too short (minimum is 5 characters)",
        "Password starts or ends with spaces.",
      ],
    });
  });
});
