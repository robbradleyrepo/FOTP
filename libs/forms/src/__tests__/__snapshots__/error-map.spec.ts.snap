// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserErrorsMapper should throw if any errors are missing a map entry 1`] = `
"Customer user errors contained unmapped fields:
  {\\"code\\":\\"CONTAINS_HTML_TAGS\\",\\"field\\":[\\"input\\",\\"firstName\\"],\\"message\\":\\"First name cannot contain HTML tags\\"}
,{\\"code\\":\\"CONTAINS_HTML_TAGS\\",\\"field\\":[\\"input\\",\\"lastName\\"],\\"message\\":\\"Last name cannot contain HTML tags\\"}
"
`;
